src/hotspot/share/logging/logDecorations.cpp

```diff
@@ -53,89 +49,91 @@ const char* LogDecorations::host_name() {
   return host_name;
 }
 
-void LogDecorations::create_decorations(const LogDecorators &decorators) {
-  char* position = _decorations_buffer;
-  #define DECORATOR(full_name, abbr) \
-  if (decorators.is_decorator(LogDecorators::full_name##_decorator)) { \
-    _decoration_offset[LogDecorators::full_name##_decorator] = position; \
-    position = create_##full_name##_decoration(position) + 1; \
-  } else { \
-    _decoration_offset[LogDecorators::full_name##_decorator] = NULL; \
-  }
+LogDecorations::LogDecorations(LogLevelType level, const LogTagSet &tagset, const LogDecorators &decorators) :
+  // When constructing the LogDecorations we resolve values for the requested decorators.
+  //
+  // _millis: needed for "time", "utctime", "timemillis":
+  _millis(
+      (decorators.is_decorator(LogDecorators::time_decorator) ||
+       decorators.is_decorator(LogDecorators::utctime_decorator) ||
+       decorators.is_decorator(LogDecorators::timemillis_decorator)) ? os::javaTimeMillis() : 0),
+  // _nanos: needed for "timenanos"
+  _nanos(decorators.is_decorator(LogDecorators::timenanos_decorator) ? os::javaTimeNanos() : 0),
+  // _elapsed_seconds: needed for "uptime", "uptimemillis", "uptimenanos"
+  _elapsed_seconds(
+      (decorators.is_decorator(LogDecorators::uptime_decorator) ||
+       decorators.is_decorator(LogDecorators::uptimemillis_decorator) ||
+       decorators.is_decorator(LogDecorators::uptimenanos_decorator)) ? os::elapsedTime() : 0),
+  // tid
+  _tid(decorators.is_decorator(LogDecorators::tid_decorator) ? os::current_thread_id() : 0),
+  // the rest is handed down by the caller
+  _level(level), _tagset(tagset)
+{
+}
+
+void LogDecorations::print_decoration(LogDecorators::Decorator decorator, outputStream* st) const {
```

Comment:
Is that a good idea to do a sanity check here in debug build? 

In my understanding, calling `print_#name##decoration()` for a decorator which was not set in ctor's decorators is undefined.  Currently, your code is safe because you carefully initialize each member variable in initializer list. You can't assume follower programmers do the same. 

If you save the variable `decorators`(a bitmask), you can do sanity check here.  This can be done in debug build. eg. 
`assert(_decorators.is_decorator(decorator), "sanity check");`

Reply:
Hmm, would not the compiler warn you if you leave member initialization out?

Your idea is not bad, but I usually avoid adding debug only members because you end up with different memory size and layout of data structures between debug and opt builds, which may have strange consequences when hunting bugs.

Idk. I can do this here, I have no strong feelings either way.

Reply:
C++ won't warn you if you don't initialize member variables. Unlike Java,  C++ even doesn't zero them for you by default.  

eg. 
```
#include <stdio.h>
class T{
public:
    int a, b, c;
    T() : b(0) {}
};

int main() {
    T t;
    printf("%d\n", t.a); // random value on stack
    printf("%d\n", t.b); // 0
    printf("%d\n", t.c); // random value on stack
}
```

It's relatively safe now because all member variables you added to LogDecorations are "scalars".  If we extend an object decorator, printing an uninitialized object may result in undefined behavior.   

I am fine with/without it. can leave it to the future.

Reply:
Okay, I'll add the check

Reply:
(I think static checkers like Sonarcloud would pick those simple errors up)